### PR TITLE
[resolution pt.2] required package variable source

### PR DESCRIPTION
**Pull request by perdasilva**
This PR adds the `RequiredPackage` variable source for generating required package variables from package names

### CODE REVIEW

**Comment by tmshort** [comment]
These all seem to build upon each other. So, part 2, includes part 1? And part 3 includes parts 1 and 2?

**Comment by dtfranz** [comment]
> These all seem to build upon each other. So, part 2, includes part 1? And part 3 includes parts 1 and 2?

Yes, I've been reviewing them individually by looking at changes from only the last commit e.g. for this one: [d9b2429](https://github.com/operator-framework/operator-controller/pull/96/commits/d9b242992ca007b220420f86236871145b3b4f09)

**Comment by tmshort** [comment]
> > These all seem to build upon each other. So, part 2, includes part 1? And part 3 includes parts 1 and 2?
> 
> Yes, I've been reviewing them individually by looking at changes from only the last commit e.g. for this one: [d9b2429](https://github.com/operator-framework/operator-controller/pull/96/commits/d9b242992ca007b220420f86236871145b3b4f09)

Because the merges aren't fast-forward, this might get strange...

File: internal/resolution/variable_sources/required_package/required_package.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	for _, bundle := range bundleEntities {
+		entityIDs = append(entityIDs, bundle.ID)
+	}
+	return &RequiredPackageVariable{
+		SimpleVariable: input.NewSimpleVariable(id, constraint.Mandatory(), constraint.Dependency(entityIDs...)),
+		bundleEntities: bundleEntities,
+	}
+}
+
+var _ input.VariableSource = &RequiredPackageVariableSource{}
```
Comment (tmshort) [inline internal/resolution/variable_sources/required_package/required_package.go:37]: Not sure I get this construct?

Comment (dtfranz) [reply internal/resolution/variable_sources/required_package/required_package.go:37]: It's ensuring that `RequiredPackageVariableSource` correctly implements `input.VariableSource`: https://go.dev/doc/effective_go#blank_implements

File: internal/resolution/variable_sources/required_package/required_package.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+type RequiredPackageVariableSource struct {
+	packageName string
+}
+
+func NewRequiredPackage(packageName string) *RequiredPackageVariableSource {
```
Comment (awgreene) [inline internal/resolution/variable_sources/required_package/required_package.go:43]: If RequiredPackageVariableSource is exported, why do I need this function?

File: internal/resolution/variable_sources/required_package/required_package.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	requiredPackage := &RequiredPackageVariableSource{
+		packageName: packageName,
+	}
+	return requiredPackage
```
Comment (awgreene) [inline internal/resolution/variable_sources/required_package/required_package.go:47]: Non-blocking nit:
```suggestion
        return &RequiredPackageVariableSource{
		packageName: packageName,
	}
```

File: internal/resolution/variable_sources/required_package/required_package_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+				property.TypePackage: `{"packageName": "test-package", "version": "2.0.0"}`,
+				property.TypeChannel: `{"channelName":"stable","priority":0}`,
+			})),
+			olmentity.NewBundleEntity(input.NewEntity("bundle-1", map[string]string{
+				property.TypePackage: `{"packageName": "test-package", "version": "1.0.0"}`,
+				property.TypeChannel: `{"channelName":"stable","priority":0}`})),
+		}))
+	})
+
+	It("should return error if package not found", func() {
```
Comment (awgreene) [inline internal/resolution/variable_sources/required_package/required_package_test.go:123]: Non-blocking nit:
```suggestion
It("should return an error if package not found", func() {
```

**Review by awgreene — APPROVED** [review]
Great work Per

File: internal/resolution/variable_sources/required_package/required_package.go (hunk 2 above)
Comment (perdasilva) [reply internal/resolution/variable_sources/required_package/required_package.go:43]: Just being more idiomatic, I suppose. Plus if something changes in the initialization we don't need to refactor every call.

File: internal/resolution/variable_sources/required_package/required_package_test.go (hunk 4 above)
Comment (perdasilva) [reply internal/resolution/variable_sources/required_package/required_package_test.go:123]: fixed ^^

**Pull request merged**